### ISSUE TITLE
Create Policy-Check API

### ISSUE THREAD

**Issue by mekarpeles** [internetarchive/openlibrary#2194]
# Background

## What is a Policy-Check API?

Sometimes libraries deem certain books ineligible for their lending programs because of  conflicts with their organizational policies. For instance, some universities may choose to only lend to their own students. Some libraries may choose not to lend audiobooks because the infrastructure isn't available to support these media format. Some specialty libraries may choose to lend works on a certain topic or in specific languages.

## Why does Open Library need a Policy-Check API?

In issue #659 we describe a program which empowers patrons to fund/sponsor books which are missing from Internet Archive's lending program. There are several criteria which must be met in order for a book to be eligible for the Book Sponsorship Program.

## What criteria is necessary for Open Library Policy-Check?

In order for us to sustainable and respectfully run a successful Book Sponsorship Program, a book must meet the following conditions to be eligible for sponsorship:

1. The price (with shipping) of the book has to be under a certain amount (e.g. < ~$20.00)
2. Due to limited physical archive space, we want the ability to prioritize unique works (i.e. works for which we have yet to obtain any editions).
3. We want to ensure the book meets all of Internet Archive's other policies.

## Where does the data for a Policy-Check come from?

Open Library will leverage 3 endpoints to perform a Policy-Check:

1. The internal `Price API`. The price restrictions are answered using the Better World Books API (which includes shipping costs in their prices), as well as the Amazon Affiliate API, both which are accessed via e.g. https://openlibrary.org/prices?isbn=9780062457714. This API is not intended for bulk use and agents which abuse this endpoint may be blacklisted from the openlibrary.org service.

2. The `Archive.org Availability API`. When answering the question "do we already have another edition of this work", we can use the Archive.org Availability API and provide and `openlibrary_work` ID: e.g. http://archive.org/services/availability?openlibrary_work=OL13716956W

3. The internal Archive.org `Policy Check` (which doesn't yet exist). This is a privileged internal API (likely a modification of @judec's "DWWI" -- http://archive.org/services/dwwi.php?barcode=xxx) which requires s3 keys of a privileged user account. This API may use heuristics and suggestions (cc: @jjjake) to weed out content which doesn't match our policy themes (e.g. potentially elicit content).

## Where will the Open Library Policy-Check API live?

1. One option is for the ~Open Library Policy-Check API to be accessed via the existing Open Library Availability `isbn` endpoint: `https://openlibrary.org/isbn/xxx`~ or
2. ~Extend The Prices `https://openlibrary.org/prices?isbn=9780062457714` endpoint to accept something like `?options=policycheck`~ or
3. ✅ create a new endpoint like `https://openlibrary.org/policy-check?isbn=0123456789`

## What parameters does Open Library Policy-Check API accept?
The Open Library Policy-Check API may be called with the following keys:
- `isbn`, e.g. `9780062457714` (v1)
- `openlibrary_edition` e.g. `OL26193423M` (v1)
- `openlibrary_work` e.g. `OL14871357W` (v2)

## What happens when the Open Library Policy-Check API is called?

If an `openlibrary_edition` is provided, the edition record will be fetched using `web.ctx.site.get('/books/:openlibrary_edition')`. If the edition has no `isbn`, the book fails our Policy-Check with a status like `no_isbn` -- we're only after books with `isbn` for this program. Otherwise, we continue with the standard `isbn` flow.

If an isbn is provided which does exist on Open Library, how do we determine if it’s eligible? If an `isbn` is provided to the API and the isbn does not exist on openlibrary.org, the endpoint will fetch and create the record using the Amazon Affiliate API to ensure its existence. Next the `isbn` will be fed into the `Archive.org Availability API`, e.g. http://archive.org/services/availability?isbn=9780062457714 and either fail the Policy-Check with a status of `work_already_available` (probably returning the info of this work so it can be displayed to the end user) or, the process proceeds to check `price` and `policy` endpoints.

### Policy-checking Works (v2)
Search results go by `work` not by edition, so there may be a need to policy-check an entire work. In v2, if an `openlibrary_work` id is provided to the Open Library `Policy-Check API`, we first do an `Archive.org Availability API` lookup (e.g. https://archive.org/services/availability?openlibrary_work=OL13716956W) and ensure we don't already have a copy of the book. If the API discovers that an edition of this work is already available to read or borrow, the work fails our Policy-Check with an error like `work_already_available`. Otherwise, we get the editions of the work, compile a set of known `isbn`s for the editions (i.e. the equivalent of http://openlibrary.org/works/OL13716956W/editions.json), and proceed to a `price` and `policy` check.

# Subtasks

1. [ ] @mekarpeles: Coordinate with @judec and @jjjake and Alexis to build in the Archive.org Policy-Check logic into DWWI? We should likely choose a different service endpoint as we'll have to pass book data **to** the service (whereas most of our existing policy-checks are on existing Archive.org items and the metadata is already available on the back-end).
2. [x] @tabshaikh : Spec out the Openlibrary.org Policy-Check end-point in another issue (what does the output data look like?)
3. [ ] @tabshaikh: Mock unit test cases with fake book data to ensure different cases (described above) correctly pass / fail policy-check.

**Comment by mekarpeles** [comment]
Example API response may look like:
```
{
  OL26193423M: {
    eligible_for_sponsorship: false,
    reasons: {
      meets_ia_policies: false,
      price: {
        betterworldbooks: {
          url: "http://www.anrdoezrs.net/links/8196539/type/dlg/http://www.betterworldbooks.com/-id-9780062457714.aspx",
          price: "$14.48 (used)",
          qlt: "used",
          price_amt: "14.48"
        }
      },
      availability: {
        status: "private",
        identifier: "isbn-9780062457714",
        isbn: "9780062457714",
        oclc: null,
        openlibrary_work: "OL17590212W", 
        openlibrary_edition: "OL26193423M",
        last_loan_date: null,
        num_waitlist: null,
        last_waitlist_date: null,
        collection: "openlibraryscanningteam,printdisabled"
      }
    }
  }
}
```

**Comment by mekarpeles** [comment]
The `policy_check` method will likely live in `plugins/openlibrary/api.py` and look something like:
```
def policy_check(isbn):
   # hits prices
   prices = get_prices(isbn)  # I think there's something like this in core/vendor.py
   # hits availability
   availability = get_availability(isbn) # this exact function is named something different in plugins/openlibrary/api.py
   # hits Archive.org policy api
   book = web.ctx.site.get(isbn)  # this isn't the exact syntax, but we need to use the isbn to get the book before we do anything else because the archive policy API needs the book's data 
   passes_policy = get_archive_policy(isbn, book.title, book.year, book.author, book.publisher)  # this function doesn't exist yet, will probably go in plugins/openlibrary/api.py -- what does this data format look like?!?!?!
  eligible_for_sponsorship = true if not prices and not available and passes_policy
   return {
     "eligible_for_sponsorship": eligible_for_sponsorship
     "reasons": {
       "prices": prices,
       "availability": availability
     }
   }
```

One of the big unknowns is what the data output looks like for the `get_archive_policy` http request.

**Comment by mekarpeles** [comment]
I talked to @samuel-archive. @jjjake has a git lab which will may help us answer some of the policy questions. It's in python.

Thus, instead of having this policy-check API on archive.org, I think it makes sense to add to openlibrary.org as `https://openlibrary.org/api/oracle`.

The way it will work (without leaking anything private) is the production service will import @jjjake's code as a module (or the `oracle` will gracefully fall back to always answering `false`) as a dependency to `openlibrary` from git lab.

cc: @bfalling

**Comment by mekarpeles** [comment]
This is merged -- the book sponsorship eligibility 3rd party library needs to be pip installed manually on the production servers.

It also still needs to be wired up from archive.org against `/sponsorship/eligibility/{isbn}` which is where this API now lives

cc: @tabshaikh

**Comment by tabshaikh** [comment]
did you replace the parameter with / ?
